ethr-did/src/main/java/me/uport/sdk/ethrdid/EthrDIDResolver.kt

```diff
@@ -169,86 +154,118 @@ open class EthrDIDResolver(
+        val matchResult = regex.matchEntire(name)
+                ?: return (pkEntries to serviceEntries)
+        val (section, algo, _, rawType, _, encoding)
+                = matchResult.destructured
+        val type = parseType(algo, rawType)
+
+        when (section) {
+
+            "pub" -> {
+                delegateIndex++
+                val pk = PublicKeyEntry(
+                        id = "$normalizedDid#delegate-$delegateIndex",
+                        type = type,
+                        owner = normalizedDid)
+
+                pkEntries[key] = when (encoding) {
+                    "", "null", "hex" ->
+                        pk.copy(publicKeyHex = event.value.items.toHexString())
+                    "base64" ->
+                        pk.copy(publicKeyBase64 = event.value.items.toBase64())
+                    "base58" ->
+                        pk.copy(publicKeyBase58 = event.value.items.toString(utf8).hexToByteArray().encodeToBase58String())
+                    else ->
+                        pk.copy(value = event.value.items.toHexString())
+                }
+
+            }
+
+            "svc" -> {
+                serviceEntries[key] = ServiceEntry(
+                        type = algo,
+                        serviceEndpoint = event.value.items.toString(utf8)
+                )
+            }
+        }
+        return (pkEntries to serviceEntries)
+    }
+
+    @Suppress("StringLiteralDuplication")
+    private fun processDelegateChanged(event: DIDDelegateChanged.Arguments, delegateCount: Int, ownerDID: String):
+            Pair<MutableMap<String, PublicKeyEntry>, MutableMap<String, AuthenticationEntry>> {
+
+        val pkEntries = mapOf<String, PublicKeyEntry>().toMutableMap()
+        val authEntries = mapOf<String, AuthenticationEntry>().toMutableMap()
+
+        var delegateIndex = delegateCount
+        val delegateType = event.delegatetype.bytes.toString(utf8)
+        val delegate = event.delegate.value.toHexStringNoPrefix().prepend0xPrefix()
+        val key = "DIDDelegateChanged-$delegateType-$delegate"
+        val validTo = event.validto.value.toLong()
+
+        if (validTo >= timeProvider.nowMs() / 1000L) {
+            delegateIndex++
+
+            when (delegateType) {
+                Secp256k1SignatureAuthentication2018.name,
+                sigAuth -> authEntries[key] = AuthenticationEntry(
+                        type = Secp256k1SignatureAuthentication2018,
+                        publicKey = "$ownerDID#delegate-$delegateIndex")
+
+                Secp256k1VerificationKey2018.name,
+                veriKey -> pkEntries[key] = PublicKeyEntry(
+                        id = "$ownerDID#delegate-$delegateIndex",
+                        type = Secp256k1VerificationKey2018,
+                        owner = ownerDID,
+                        ethereumAddress = delegate)
+            }
+        }
+        return (pkEntries to authEntries)
+    }
+
```

Comment:
Good call on breaking things into smaller functions

Reply:
Yeah, I wish I had found a better way though because this involves some extra allocations.
Even so, I think it makes the code more readable.